File: src/core/devices/primitives.js
/**
 * Linear primitives: R, C, L, K, and the four controlled sources.
 *
 * These carry the stamp conventions everything else follows:
 *   nodes[0] = positive terminal, current flows in at nodes[0].
 *   branches[] holds indices of extra unknowns owned by this device.
 */
import { Device } from '../device.js';

/** Two-terminal conductance stamp handles. */
function reserve2(sys, p, n) {
  sys.reserve(p, p); sys.reserve(p, n); sys.reserve(n, p); sys.reserve(n, n);
}
function bind2(sys, p, n) {
  return [sys.handle(p, p), sys.handle(p, n), sys.handle(n, p), sys.handle(n, n)];
}
function stamp2(sys, h, g) {
  sys.add(h[0], g); sys.add(h[1], -g); sys.add(h[2], -g); sys.add(h[3], g);
}
function stamp2c(sys, h, gr, gi) {
  sys.addComplex(h[0], gr, gi); sys.addComplex(h[1], -gr, -gi);
  sys.addComplex(h[2], -gr, -gi); sys.addComplex(h[3], gr, gi);
}

export class Resistor extends Device {
  constructor(name, p, n, r) {
    super(name);
    this.nodes = [p, n];
    this.r = r;
    this.tc1 = 0;
    this.tc2 = 0;
    this.g = 1 / r;
  }
  temperature(ctx) {
    const dt = ctx.temp - ctx.nomTemp;
    const rEff = this.r * (1 + this.tc1 * dt + this.tc2 * dt * dt);
    this.g = 1 / rEff;
  }
  reserve(sys) { reserve2(sys, this.nodes[0], this.nodes[1]); }
  bind(sys) { this.h = bind2(sys, this.nodes[0], this.nodes[1]); }
  loadDC(ctx) { stamp2(ctx.sys, this.h, this.g); }
  loadNoise(ctx, out) {
    // Johnson noise: 4kTG A^2/Hz, injected between the terminals.
    out.push({ device: this.name, type: 'thermal', nodes: this.nodes, psd: 4 * 1.380649e-23 * ctx.temp * this.g });
  }
}

export class Capacitor extends Device {
  constructor(name, p, n, c) {
    super(name);
    this.nodes = [p, n];
    this.c = c;
    this.ic = null;      // initial condition, applied when uic is set
    this.nStates = 2;    // [charge, current]
  }
  reserve(sys) { reserve2(sys, this.nodes[0], this.nodes[1]); }
  bind(sys) { this.h = bind2(sys, this.nodes[0], this.nodes[1]); }

  loadDC(ctx) {
    // Open circuit at DC. gmin across the branch keeps the matrix nonsingular
    // when a capacitor is the only element on a node.
    stamp2(ctx.sys, this.h, ctx.gmin);
  }

  loadTran(ctx) {
    const [p, n] = this.nodes;
    const v = ctx.v(p) - ctx.v(n);
    const st = ctx.state.at(0);
    st[this.stateOff] = this.c * v;
    ctx.integrate(this.stateOff, this.c, v);
    stamp2(ctx.sys, this.h, ctx.geq);
    ctx.rhs(p, -ctx.ieq);
    ctx.rhs(n, ctx.ieq);
  }

  loadAC(ctx) {
    stamp2c(ctx.sys, this.h, 0, ctx.omega * this.c);
  }

  maxTimestep(ctx) {
    return this._lte === undefined ? Infinity : this._lte;
  }
}

export class Inductor extends Device {
  constructor(name, p, n, l) {
    super(name);
    this.nodes = [p, n];
    this.l = l;
    this.ic = null;
    this.nBranches = 1;
    this.nStates = 2;    // [flux, voltage]
    /** @type {Inductor[]} mutual couplings, filled in by Coupling */
    this.couples = [];
  }
  reserve(sys) {
    const [p, n] = this.nodes;
    const b = this.branches[0];
    sys.reserve(p, b); sys.reserve(n, b);
    sys.reserve(b, p); sys.reserve(b, n); sys.reserve(b, b);
    for (const { other } of this.couples) sys.reserve(b, other.branches[0]);
  }
  bind(sys) {
    const [p, n] = this.nodes;
    const b = this.branches[0];
    this.hPB = sys.handle(p, b);
    this.hNB = sys.handle(n, b);
    this.hBP = sys.handle(b, p);
    this.hBN = sys.handle(b, n);
    this.hBB = sys.handle(b, b);
    this.hMut = this.couples.map((c) => sys.handle(b, c.other.branches[0]));
  }
  _stampTopology(sys) {
    sys.add(this.hPB, 1);
    sys.add(this.hNB, -1);
    sys.add(this.hBP, 1);
    sys.add(this.hBN, -1);
  }
  loadDC(ctx) {
    // A short at DC: V(p) - V(n) = 0.
    this._stampTopology(ctx.sys);
  }
  loadTran(ctx) {
    const b = this.branches[0];
    const i = ctx.x[b];
    const st = ctx.state.at(0);
    let flux = this.l * i;
    for (const c of this.couples) flux += c.m * ctx.x[c.other.branches[0]];
    st[this.stateOff] = flux;
    ctx.integrate(this.stateOff, this.l, i);
    this._stampTopology(ctx.sys);
    ctx.sys.add(this.hBB, -ctx.geq);
    // ctx.ieq already folds in the mutual flux through the charge state, so the
    // linear mutual term is subtracted out of the RHS as it is added to the
    // matrix. scale converts dq/dv=L into the method's coefficient (L/dt for BE).
    let ieq = ctx.ieq;
    const scale = ctx.geq / this.l;
    this.couples.forEach((c, k) => {
      const gm = scale * c.m;
      ctx.sys.add(this.hMut[k], -gm);
      ieq -= gm * ctx.x[c.other.branches[0]];
    });
    ctx.rhs(b, ieq);
  }
  loadAC(ctx) {
    this._stampTopology(ctx.sys);
    ctx.sys.addComplex(this.hBB, 0, -ctx.omega * this.l);
    this.couples.forEach((c, k) => {
      ctx.sys.addComplex(this.hMut[k], 0, -ctx.omega * c.m);
    });
  }
}

/** Mutual inductance: K<name> L1 L2 <coupling coefficient>. */
export class Coupling extends Device {
  constructor(name, l1, l2, k) {
    super(name);
    this.l1 = l1;
    this.l2 = l2;
    this.k = k;
    const m = k * Math.sqrt(l1.l * l2.l);
    this.m = m;
    l1.couples.push({ other: l2, m });
    l2.couples.push({ other: l1, m });
  }
}

export class VCVS extends Device {
  constructor(name, p, n, cp, cn, gain) {
    super(name);
    this.nodes = [p, n, cp, cn];
    this.gain = gain;
    this.nBranches = 1;
  }
  reserve(sys) {
    const [p, n, cp, cn] = this.nodes;
    const b = this.branches[0];
    sys.reserve(p, b); sys.reserve(n, b);
    sys.reserve(b, p); sys.reserve(b, n); sys.reserve(b, cp); sys.reserve(b, cn);
  }
  bind(sys) {
    const [p, n, cp, cn] = this.nodes;
    const b = this.branches[0];
    this.h = [sys.handle(p, b), sys.handle(n, b), sys.handle(b, p),
              sys.handle(b, n), sys.handle(b, cp), sys.handle(b, cn)];
  }
  loadDC(ctx) {
    const s = ctx.sys;
    s.add(this.h[0], 1); s.add(this.h[1], -1);
    s.add(this.h[2], 1); s.add(this.h[3], -1);
    s.add(this.h[4], -this.gain); s.add(this.h[5], this.gain);
  }
}

export class VCCS extends Device {
  constructor(name, p, n, cp, cn, gm) {
    super(name);
    this.nodes = [p, n, cp, cn];
    this.gm = gm;
  }
  reserve(sys) {
    const [p, n, cp, cn] = this.nodes;
    sys.reserve(p, cp); sys.reserve(p, cn); sys.reserve(n, cp); sys.reserve(n, cn);
  }
  bind(sys) {
    const [p, n, cp, cn] = this.nodes;
    this.h = [sys.handle(p, cp), sys.handle(p, cn), sys.handle(n, cp), sys.handle(n, cn)];
  }
  loadDC(ctx) {
    const s = ctx.sys;
    s.add(this.h[0], this.gm); s.add(this.h[1], -this.gm);
    s.add(this.h[2], -this.gm); s.add(this.h[3], this.gm);
  }
}

/** Current-controlled current source. Control is the branch of a voltage source. */
export class CCCS extends Device {
  constructor(name, p, n, ctrl, gain) {
    super(name);
    this.nodes = [p, n];
    this.ctrl = ctrl;
    this.gain = gain;
  }
  reserve(sys) {
    const cb = this.ctrl.branches[0];
    sys.reserve(this.nodes[0], cb);
    sys.reserve(this.nodes[1], cb);
  }
  bind(sys) {
    const cb = this.ctrl.branches[0];
    this.h = [sys.handle(this.nodes[0], cb), sys.handle(this.nodes[1], cb)];
  }
  loadDC(ctx) {
    ctx.sys.add(this.h[0], this.gain);
    ctx.sys.add(this.h[1], -this.gain);
  }
}

/** Current-controlled voltage source. */
export class CCVS extends Device {
  constructor(name, p, n, ctrl, gain) {
    super(name);
    this.nodes = [p, n];
    this.ctrl = ctrl;
    this.gain = gain;
    this.nBranches = 1;
  }
  reserve(sys) {
    const [p, n] = this.nodes;
    const b = this.branches[0];
    const cb = this.ctrl.branches[0];
    sys.reserve(p, b); sys.reserve(n, b);
    sys.reserve(b, p); sys.reserve(b, n); sys.reserve(b, cb);
  }
  bind(sys) {
    const [p, n] = this.nodes;
    const b = this.branches[0];
    this.h = [sys.handle(p, b), sys.handle(n, b), sys.handle(b, p),
              sys.handle(b, n), sys.handle(b, this.ctrl.branches[0])];
  }
  loadDC(ctx) {
    const s = ctx.sys;
    s.add(this.h[0], 1); s.add(this.h[1], -1);
    s.add(this.h[2], 1); s.add(this.h[3], -1);
    s.add(this.h[4], -this.gain);
  }
}

/** Voltage-controlled switch with hysteresis. */
export class VSwitch extends Device {
  constructor(name, p, n, cp, cn, model) {
    super(name);
    this.nodes = [p, n, cp, cn];
    this.m = { vt: 0, vh: 0, ron: 1, roff: 1e12, ...model };
    this.on = false;
    this.nonlinear = true;
  }
  reserve(sys) { reserve2(sys, this.nodes[0], this.nodes[1]); }
  bind(sys) { this.h = bind2(sys, this.nodes[0], this.nodes[1]); }
  loadDC(ctx) {
    const vc = ctx.v(this.nodes[2]) - ctx.v(this.nodes[3]);
    const { vt, vh, ron, roff } = this.m;
    if (vc > vt + vh) this.on = true;
    else if (vc < vt - vh) this.on = false;
    stamp2(ctx.sys, this.h, 1 / (this.on ? ron : roff));
  }
}

File: crates/spicelab-core/src/devices/primitives.rs
//! Linear primitives: R, C, L, K, and the four controlled sources.
//! Port of `src/core/devices/primitives.js`.
//!
//! Stamp conventions everything else follows:
//!   `nodes[0]` = positive terminal, current flows in at `nodes[0]`.
//!   `branches[]` holds indices of extra unknowns owned by this device.

use crate::context::Context;
use crate::device::{bind2, reserve2, stamp2, stamp2c, Common, DeviceOps, NoiseSource};
use crate::sparse::SparseSystem;

const BOLTZMANN: f64 = 1.380649e-23;

pub struct Resistor {
    pub c: Common,
    pub r: f64,
    pub tc1: f64,
    pub tc2: f64,
    g: f64,
    h: [i32; 4],
}

impl Resistor {
    pub fn new(name: &str, p: i32, n: i32, r: f64) -> Self {
        Resistor {
            c: Common::new(name, vec![p, n]),
            r,
            tc1: 0.0,
            tc2: 0.0,
            g: 1.0 / r,
            h: [-1; 4],
        }
    }
}

impl DeviceOps for Resistor {
    fn common(&self) -> &Common {
        &self.c
    }
    fn common_mut(&mut self) -> &mut Common {
        &mut self.c
    }

    fn temperature(&mut self, ctx: &Context) {
        let dt = ctx.temp - ctx.nom_temp;
        let r_eff = self.r * (1.0 + self.tc1 * dt + self.tc2 * dt * dt);
        self.g = 1.0 / r_eff;
    }

    fn reserve(&self, sys: &mut SparseSystem) {
        reserve2(sys, self.c.nodes[0], self.c.nodes[1]);
    }
    fn bind(&mut self, sys: &SparseSystem) {
        self.h = bind2(sys, self.c.nodes[0], self.c.nodes[1]);
    }
    fn load_dc(&mut self, ctx: &mut Context) {
        stamp2(&mut ctx.sys, &self.h, self.g);
    }

    fn load_noise(&self, ctx: &Context, out: &mut Vec<NoiseSource>) {
        // Johnson noise: 4kTG A^2/Hz, injected between the terminals.
        out.push(NoiseSource {
            device: self.c.name.clone(),
            kind: "thermal",
            nodes: [self.c.nodes[0], self.c.nodes[1]],
            psd: 4.0 * BOLTZMANN * ctx.temp * self.g,
        });
    }
}

pub struct Capacitor {
    pub c: Common,
    pub cap: f64,
    /// Initial condition, applied when `uic` is set.
    pub ic: Option<f64>,
    h: [i32; 4],
}

impl Capacitor {
    pub fn new(name: &str, p: i32, n: i32, cap: f64) -> Self {
        Capacitor {
            c: Common::new(name, vec![p, n]),
            cap,
            ic: None,
            h: [-1; 4],
        }
    }
}

impl DeviceOps for Capacitor {
    fn common(&self) -> &Common {
        &self.c
    }
    fn common_mut(&mut self) -> &mut Common {
        &mut self.c
    }
    /// [charge, current]
    fn n_states(&self) -> usize {
        2
    }

    fn reserve(&self, sys: &mut SparseSystem) {
        reserve2(sys, self.c.nodes[0], self.c.nodes[1]);
    }
    fn bind(&mut self, sys: &SparseSystem) {
        self.h = bind2(sys, self.c.nodes[0], self.c.nodes[1]);
    }

    fn load_dc(&mut self, ctx: &mut Context) {
        // Open circuit at DC. gmin across the branch keeps the matrix
        // nonsingular when a capacitor is the only element on a node.
        let g = ctx.gmin;
        stamp2(&mut ctx.sys, &self.h, g);
    }

    fn load_tran(&mut self, ctx: &mut Context) {
        let (p, n) = (self.c.nodes[0], self.c.nodes[1]);
        let v = ctx.v(p) - ctx.v(n);
        let off = self.c.state_off;
        ctx.state.at_mut(0)[off] = self.cap * v;
        ctx.integrate(off, self.cap, v);
        let (geq, ieq) = (ctx.geq, ctx.ieq);
        stamp2(&mut ctx.sys, &self.h, geq);
        ctx.rhs(p, -ieq);
        ctx.rhs(n, ieq);
    }

    fn load_ac(&mut self, ctx: &mut Context) {
        let b = ctx.omega * self.cap;
        stamp2c(&mut ctx.sys, &self.h, 0.0, b);
    }
}

/// A mutual-inductance coupling, resolved to a concrete branch index during
/// topology build. `K<name> L1 L2 <coefficient>` creates one of these on each
/// of the two inductors.
#[derive(Clone)]
pub struct Couple {
    /// Index of the coupled inductor in `Circuit::devices`.
    pub other_dev: usize,
    /// Mutual inductance, `k * sqrt(l1 * l2)`.
    pub m: f64,
    /// Branch index of the coupled inductor. Filled in by `build_topology`.
    pub other_branch: usize,
}

pub struct Inductor {
    pub c: Common,
    pub l: f64,
    pub ic: Option<f64>,
    pub couples: Vec<Couple>,
    h_pb: i32,
    h_nb: i32,
    h_bp: i32,
    h_bn: i32,
    h_bb: i32,
    h_mut: Vec<i32>,
}

impl Inductor {
    pub fn new(name: &str, p: i32, n: i32, l: f64) -> Self {
        Inductor {
            c: Common::new(name, vec![p, n]),
            l,
            ic: None,
            couples: Vec::new(),
            h_pb: -1,
            h_nb: -1,
            h_bp: -1,
            h_bn: -1,
            h_bb: -1,
            h_mut: Vec::new(),
        }
    }

    fn stamp_topology(&self, sys: &mut SparseSystem) {
        sys.add(self.h_pb, 1.0);
        sys.add(self.h_nb, -1.0);
        sys.add(self.h_bp, 1.0);
        sys.add(self.h_bn, -1.0);
    }
}

impl DeviceOps for Inductor {
    fn common(&self) -> &Common {
        &self.c
    }
    fn common_mut(&mut self) -> &mut Common {
        &mut self.c
    }
    fn n_branches(&self) -> usize {
        1
    }
    /// [flux, voltage]
    fn n_states(&self) -> usize {
        2
    }

    fn reserve(&self, sys: &mut SparseSystem) {
        let (p, n) = (self.c.nodes[0], self.c.nodes[1]);
        let b = self.c.branches[0] as i32;
        sys.reserve(p, b);
        sys.reserve(n, b);
        sys.reserve(b, p);
        sys.reserve(b, n);
        sys.reserve(b, b);
        for c in &self.couples {
            sys.reserve(b, c.other_branch as i32);
        }
    }

    fn bind(&mut self, sys: &SparseSystem) {
        let (p, n) = (self.c.nodes[0], self.c.nodes[1]);
        let b = self.c.branches[0] as i32;
        self.h_pb = sys.handle(p, b);
        self.h_nb = sys.handle(n, b);
        self.h_bp = sys.handle(b, p);
        self.h_bn = sys.handle(b, n);
        self.h_bb = sys.handle(b, b);
        self.h_mut = self
            .couples
            .iter()
            .map(|c| sys.handle(b, c.other_branch as i32))
            .collect();
    }

    fn load_dc(&mut self, ctx: &mut Context) {
        // A short at DC: V(p) - V(n) = 0.
        self.stamp_topology(&mut ctx.sys);
    }

    fn load_tran(&mut self, ctx: &mut Context) {
        let b = self.c.branches[0];
        let i = ctx.x[b];
        let mut flux = self.l * i;
        for c in &self.couples {
            flux += c.m * ctx.x[c.other_branch];
        }
        let off = self.c.state_off;
        ctx.state.at_mut(0)[off] = flux;
        ctx.integrate(off, self.l, i);

        self.stamp_topology(&mut ctx.sys);
        let (geq, mut ieq) = (ctx.geq, ctx.ieq);
        ctx.sys.add(self.h_bb, -geq);

        // ctx.ieq already folds in the mutual flux through the charge state, so
        // the linear mutual term is subtracted out of the RHS as it is added to
        // the matrix. `scale` converts dq/dv = L into the method's coefficient
        // (L/dt for BE).
        let scale = geq / self.l;
        for (k, c) in self.couples.iter().enumerate() {
            let gm = scale * c.m;
            ctx.sys.add(self.h_mut[k], -gm);
            ieq -= gm * ctx.x[c.other_branch];
        }
        ctx.rhs(b as i32, ieq);
    }

    fn load_ac(&mut self, ctx: &mut Context) {
        self.stamp_topology(&mut ctx.sys);
        let w = ctx.omega;
        ctx.sys.add_complex(self.h_bb, 0.0, -w * self.l);
        for (k, c) in self.couples.iter().enumerate() {
            ctx.sys.add_complex(self.h_mut[k], 0.0, -w * c.m);
        }
    }
}

pub struct Vcvs {
    pub c: Common,
    pub gain: f64,
    h: [i32; 6],
}

impl Vcvs {
    pub fn new(name: &str, p: i32, n: i32, cp: i32, cn: i32, gain: f64) -> Self {
        Vcvs {
            c: Common::new(name, vec![p, n, cp, cn]),
            gain,
            h: [-1; 6],
        }
    }
}

impl DeviceOps for Vcvs {
    fn common(&self) -> &Common {
        &self.c
    }
    fn common_mut(&mut self) -> &mut Common {
        &mut self.c
    }
    fn n_branches(&self) -> usize {
        1
    }

    fn reserve(&self, sys: &mut SparseSystem) {
        let (p, n, cp, cn) = (
            self.c.nodes[0],
            self.c.nodes[1],
            self.c.nodes[2],
            self.c.nodes[3],
        );
        let b = self.c.branches[0] as i32;
        sys.reserve(p, b);
        sys.reserve(n, b);
        sys.reserve(b, p);
        sys.reserve(b, n);
        sys.reserve(b, cp);
        sys.reserve(b, cn);
    }

    fn bind(&mut self, sys: &SparseSystem) {
        let (p, n, cp, cn) = (
            self.c.nodes[0],
            self.c.nodes[1],
            self.c.nodes[2],
            self.c.nodes[3],
        );
        let b = self.c.branches[0] as i32;
        self.h = [
            sys.handle(p, b),
            sys.handle(n, b),
            sys.handle(b, p),
            sys.handle(b, n),
            sys.handle(b, cp),
            sys.handle(b, cn),
        ];
    }

    fn load_dc(&mut self, ctx: &mut Context) {
        let s = &mut ctx.sys;
        s.add(self.h[0], 1.0);
        s.add(self.h[1], -1.0);
        s.add(self.h[2], 1.0);
        s.add(self.h[3], -1.0);
        s.add(self.h[4], -self.gain);
        s.add(self.h[5], self.gain);
    }
}

pub struct Vccs {
    pub c: Common,
    pub gm: f64,
    h: [i32; 4],
}

impl Vccs {
    pub fn new(name: &str, p: i32, n: i32, cp: i32, cn: i32, gm: f64) -> Self {
        Vccs {
            c: Common::new(name, vec![p, n, cp, cn]),
            gm,
            h: [-1; 4],
        }
    }
}

impl DeviceOps for Vccs {
    fn common(&self) -> &Common {
        &self.c
    }
    fn common_mut(&mut self) -> &mut Common {
        &mut self.c
    }

    fn reserve(&self, sys: &mut SparseSystem) {
        let (p, n, cp, cn) = (
            self.c.nodes[0],
            self.c.nodes[1],
            self.c.nodes[2],
            self.c.nodes[3],
        );
        sys.reserve(p, cp);
        sys.reserve(p, cn);
        sys.reserve(n, cp);
        sys.reserve(n, cn);
    }

    fn bind(&mut self, sys: &SparseSystem) {
        let (p, n, cp, cn) = (
            self.c.nodes[0],
            self.c.nodes[1],
            self.c.nodes[2],
            self.c.nodes[3],
        );
        self.h = [
            sys.handle(p, cp),
            sys.handle(p, cn),
            sys.handle(n, cp),
            sys.handle(n, cn),
        ];
    }

    fn load_dc(&mut self, ctx: &mut Context) {
        let s = &mut ctx.sys;
        s.add(self.h[0], self.gm);
        s.add(self.h[1], -self.gm);
        s.add(self.h[2], -self.gm);
        s.add(self.h[3], self.gm);
    }
}

/// Current-controlled current source. Control is the branch of a voltage source.
pub struct Cccs {
    pub c: Common,
    pub gain: f64,
    /// Index of the controlling device in `Circuit::devices`.
    pub ctrl_dev: usize,
    /// Branch index of the controlling device, resolved by `build_topology`.
    pub ctrl_branch: usize,
    h: [i32; 2],
}

impl Cccs {
    pub fn new(name: &str, p: i32, n: i32, ctrl_dev: usize, gain: f64) -> Self {
        Cccs {
            c: Common::new(name, vec![p, n]),
            gain,
            ctrl_dev,
            ctrl_branch: 0,
            h: [-1; 2],
        }
    }
}

impl DeviceOps for Cccs {
    fn common(&self) -> &Common {
        &self.c
    }
    fn common_mut(&mut self) -> &mut Common {
        &mut self.c
    }

    fn reserve(&self, sys: &mut SparseSystem) {
        let cb = self.ctrl_branch as i32;
        sys.reserve(self.c.nodes[0], cb);
        sys.reserve(self.c.nodes[1], cb);
    }

    fn bind(&mut self, sys: &SparseSystem) {
        let cb = self.ctrl_branch as i32;
        self.h = [sys.handle(self.c.nodes[0], cb), sys.handle(self.c.nodes[1], cb)];
    }

    fn load_dc(&mut self, ctx: &mut Context) {
        ctx.sys.add(self.h[0], self.gain);
        ctx.sys.add(self.h[1], -self.gain);
    }
}

/// Current-controlled voltage source.
pub struct Ccvs {
    pub c: Common,
    pub gain: f64,
    pub ctrl_dev: usize,
    pub ctrl_branch: usize,
    h: [i32; 5],
}

impl Ccvs {
    pub fn new(name: &str, p: i32, n: i32, ctrl_dev: usize, gain: f64) -> Self {
        Ccvs {
            c: Common::new(name, vec![p, n]),
            gain,
            ctrl_dev,
            ctrl_branch: 0,
            h: [-1; 5],
        }
    }
}

impl DeviceOps for Ccvs {
    fn common(&self) -> &Common {
        &self.c
    }
    fn common_mut(&mut self) -> &mut Common {
        &mut self.c
    }
    fn n_branches(&self) -> usize {
        1
    }

    fn reserve(&self, sys: &mut SparseSystem) {
        let (p, n) = (self.c.nodes[0], self.c.nodes[1]);
        let b = self.c.branches[0] as i32;
        let cb = self.ctrl_branch as i32;
        sys.reserve(p, b);
        sys.reserve(n, b);
        sys.reserve(b, p);
        sys.reserve(b, n);
        sys.reserve(b, cb);
    }

    fn bind(&mut self, sys: &SparseSystem) {
        let (p, n) = (self.c.nodes[0], self.c.nodes[1]);
        let b = self.c.branches[0] as i32;
        self.h = [
            sys.handle(p, b),
            sys.handle(n, b),
            sys.handle(b, p),
            sys.handle(b, n),
            sys.handle(b, self.ctrl_branch as i32),
        ];
    }

    fn load_dc(&mut self, ctx: &mut Context) {
        let s = &mut ctx.sys;
        s.add(self.h[0], 1.0);
        s.add(self.h[1], -1.0);
        s.add(self.h[2], 1.0);
        s.add(self.h[3], -1.0);
        s.add(self.h[4], -self.gain);
    }
}

#[derive(Clone, Debug)]
pub struct SwitchModel {
    pub vt: f64,
    pub vh: f64,
    pub ron: f64,
    pub roff: f64,
}

impl Default for SwitchModel {
    fn default() -> Self {
        SwitchModel {
            vt: 0.0,
            vh: 0.0,
            ron: 1.0,
            roff: 1e12,
        }
    }
}

/// Voltage-controlled switch with hysteresis.
///
/// Both switch states stamp the same two-terminal pattern, so the reserved
/// pattern is already the union and a state flip never forces re-analysis.
pub struct VSwitch {
    pub c: Common,
    pub m: SwitchModel,
    pub on: bool,
    h: [i32; 4],
}

impl VSwitch {
    pub fn new(name: &str, p: i32, n: i32, cp: i32, cn: i32, m: SwitchModel) -> Self {
        VSwitch {
            c: Common::new(name, vec![p, n, cp, cn]),
            m,
            on: false,
            h: [-1; 4],
        }
    }
}

impl DeviceOps for VSwitch {
    fn common(&self) -> &Common {
        &self.c
    }
    fn common_mut(&mut self) -> &mut Common {
        &mut self.c
    }
    fn is_nonlinear(&self) -> bool {
        true
    }

    fn reserve(&self, sys: &mut SparseSystem) {
        reserve2(sys, self.c.nodes[0], self.c.nodes[1]);
    }
    fn bind(&mut self, sys: &SparseSystem) {
        self.h = bind2(sys, self.c.nodes[0], self.c.nodes[1]);
    }

    fn load_dc(&mut self, ctx: &mut Context) {
        let vc = ctx.v(self.c.nodes[2]) - ctx.v(self.c.nodes[3]);
        if vc > self.m.vt + self.m.vh {
            self.on = true;
        } else if vc < self.m.vt - self.m.vh {
            self.on = false;
        }
        let g = 1.0 / if self.on { self.m.ron } else { self.m.roff };
        stamp2(&mut ctx.sys, &self.h, g);
    }
}
